[Wei Jue] iP

weijuey opened the pull request:
## Duke (Tentative)

Duke will be continually ~~worked on~~ improved.

Much to do (*too much* :rofl: )

1. GUI
2. More commands
   - find by date
   - set priority
3. History

> It always seems impossible until it’s done. - Nelson Mandela

Duke's good for
1. [x] Looking **good**
2. [ ] Staying in the know
3. [ ] Achieving harmony
4. [ ] having someone to talk to ([Discord](https://discord.com/) doesn't talk to you)

Seriously, look at my `main` method:

```java
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.run()
}
```
***too clean***

Commit:
Add Gradle support

weijuey committed:
Added new welcome and farewell messages as class-level members

weijuey committed:
Added BufferedReader as way of reading from command line

weijuey committed:
Changed welcome message to new welcome message, enabled chat functionality, implemented echoing and "bye" command

weijuey committed:
Fixed string matching not working, update system exit status to 0

weijuey committed:
Changed string matching to be case insensitive

weijuey committed:
Changed text given to chatbot to be stored

weijuey committed:
Added "list"command that lists all stored text

weijuey committed:
Added Task class for later implementation of to-do list

weijuey committed:
Upgraded storing text to storing tasks

weijuey committed:
Rework chatbot to pick up first word before whitespace and use it as the command word

weijuey committed:
Rework command recognition to check if command has arguments

weijuey committed:
Added functionality to mark a task not done

weijuey committed:
Added mark and unmark command to chatbot.

weijuey committed:
Reworked chatbot to not accept invalid commands, adding tasks will be readded later with new functions

weijuey committed:
Updated Task toString method, prepared classes that extend Task

weijuey committed:
Implemented ToDo task type

weijuey committed:
Implemented Event task type.

weijuey committed:
Implemented Deadline task type

weijuey committed:
Setup event, todo, deadline commands to add the respective tasks, implemented todo command

weijuey committed:
Implemented event and deadline commands

weijuey committed:
Created short sequence of test cases up to Level-4 functionalities

weijuey committed:
Updated deadline and event commands to check if deadline and duration are provided respectively

weijuey committed:
Rework DS for storing tasks to Arraylist to facilitate deletion function (to be added)

weijuey committed:
Added utility function to check validity of description, defined as having at least 1 English character

weijuey committed:
Rework event creation commands to check if given descriptor is valid

weijuey committed:
Added delete function.

weijuey committed:
Include reading a saved.txt file to startup process

weijuey committed:
Modify Deadline class to use LocalDate class to represent deadline

weijuey committed:
Change deadline command to use LocalDate.now method to create Deadline objects for now before parsing inputs for dates is added.

weijuey committed:
Add date parsing to deadline command and update deadline's string format

weijuey committed:
Change deadline to include time

weijuey committed:
Change event to also use LocalDateTime class to represent date and time of event

weijuey committed:
Add loading saved tasks on disk to chatbot startup

weijuey committed:
Add updating saved tasks to mark and unmark command when executed

weijuey committed:
Add updating saved tasks to delete command when executed

weijuey committed:
Add updating saved tasks in data to task creation commands when executed

weijuey committed:
Edit welcome and farewell messages

weijuey committed:
Merge branch 'branch-Level-7'

weijuey committed:
Merge branch 'branch-level-8'

weijuey committed:
Extract functionalities of chatbot into new class files

weijuey committed:
Place all classes into a Duke package

weijuey committed:
Add new constructor to Tasks so that they can be instantiated as completed

weijuey committed:
Extract command execution from parser into command classes

weijuey committed:
Add junit test classes for some methods of Task and Parser

weijuey committed:
Rename duke package and a few variables to follow coding convention

weijuey committed:
Correct indentation for switch statement

weijuey committed:
Add find command that lists all tasks which contains a given string

weijuey committed:
Add javadoc describing the task classes

weijuey committed:
Add javadoc describing the Duke commands

weijuey committed:
Add javadoc to the rest of the classes

weijuey committed:
Add file to .gitignore

weijuey committed:
Merge branch 'branch-A-CodingStandard'

weijuey committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/FindCommand.java

weijuey committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Duke/Parser.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Task.java

weijuey committed:
Fix find command not being recognised by parser and other merge related bugs

tandeshao commented on src/main/java/duke/Parser.java:
```diff
@@ -26,10 +26,10 @@ public Command parse(String line) throws CommandParseException {
-                    return new ByeCommand();
-                case "list":
-                    return new ListCommand();
+            case "bye":
+                return new ByeCommand();
+            case "list":
+                return new ListCommand();
```
You might want to look into using Intellij's enhanced switch-case statements. For example:
```java
switch (itemCode) {
    case 001 -> System.out.println("It's a laptop!");
    case 002 -> System.out.println("It's a desktop!");
    case 003,004 -> System.out.println("It's a mobile phone!");
}
```

tandeshao commented on src/main/java/duke/Parser.java:
```diff
@@ -34,75 +34,74 @@ public Command parse(String line) throws CommandParseException {
+                    return new DeleteCommand(taskIndex);
+                } catch (NumberFormatException e) {
+                    throw new CommandParseException("You did not provide a number!", line);
+                }
```
I like how you are following the coding standard for the switch-case statements 😄 .

tandeshao commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,21 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class ParserTest {
+    Parser parser = new Parser();
+
+    @Test
+    public void unmarkTask_invalidNumber_exceptionThrown() {
+        try {
+            assertEquals(new UnmarkCommand(0), parser.parse("unmark x"));
+            fail();
+        } catch (CommandParseException e) {
+            assertEquals("You did not provide a number!:\nunmark x", e.getMessage());
+        }
+    }
+}
```
I like how you are diligently creating test classes to test your components.

tandeshao commented on src/main/java/duke/ByeCommand.java:
```diff
@@ -1,5 +1,8 @@
 package duke;
 
+/**
+ * Duke Command for exiting the program.
+ */
```
I notice you have created java docs for all of your classes. Keep up the good work! 👍

tandeshao reviewed (commented):
LGTM

ngjunkang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,151 @@
+                case 'E':
+                    lst.addTask(new Event(des, LocalDateTime.parse(da, dateTimeParser), done == '1'));
+                    break;
+            }
```
Perhaps you can make this switch statement compliant to the Java coding standard (case indentations)? e.g.
```java
switch (type) {
case 'T':
    break;
case 'D':
    break;
}
```

On a side note, a default case can be added to detect that the file format is invalid.

ngjunkang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,151 @@
+            while (newLine != null) {
+                char[] toEdit = newLine.toCharArray();
+                if (i == index) {
+                    toEdit[1] = (toEdit[1] == '0' ? '1' : '0');
```
Maybe you might want to change this line to `toEdit[1] = (toEdit[1] == '0') ? '1' : '0';`? According to [Oracle](https://www.oracle.com/technetwork/java/codeconventions-150003.pdf) Chapter 10.5.3, it recommended us to parenthesise the expression if there's a binary operator (`==`, `<=` etc).

ngjunkang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+    private Storage savedTasks;
+
+    public Duke() {
+        ui = new Ui();
+        parser = new Parser();
+        storedTasks = new TaskList();
+        savedTasks = new Storage();
```
I feel that consistency (there's an inconsistency in the use of `this.` in the Java files) is important in a code base, what do you think?

ngjunkang commented on src/main/java/duke/EventCommand.java:
```diff
@@ -0,0 +1,27 @@
+        this.descriptor = descriptor;
+        this.dateTime = dateTime;
+    }
+
```
Perhaps the use of `@Override` annotation can be consistent across all the files too (above `execute(...)` etc)? [Google](https://google.github.io/styleguide/javaguide.html#s5-naming) recommends us to always use `@Override`.

ngjunkang commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,11 @@
+ * tasks saved to disk in some way, these behaviours are to be
+ * implemented by subclasses of this class.
+ */
+public abstract class Command {
```
I feel that all the `Command`-related classes can be put in one package, called commands, what do you think? Same for other Java files, such as `Task` and exceptions.

ngjunkang reviewed (commented):
LGTM, just some nits to fix, good job!

weijuey committed:
Update Gradle build to include checkstyle task.

weijuey committed:
Fix gradle build task not working

weijuey committed:
Move commands to a commands package and improve some javadocs and names

weijuey committed:
Improve more javadocs and names

weijuey committed:
Add a GUI and integrate Duke to it

weijuey committed:
Update GUI and fix style

weijuey committed:
Implement the rest of the GUI

weijuey committed:
Merge branch 'branch-level-10'

weijuey committed:
Add a few assert statements

weijuey committed:
Add assert statements to Storage and TaskList to ensure correctness of assumptions at points of execution

The Storage class checks that the save file for tasks when it is initialised, creating it if it doesn't. From there, the file's existence is no longer checked for. To check that the assumption that the file will continue to exist throughout the program's runtime is true, let's add assert statements to each method of Storage.

The TaskList class contains a few methods that operates on a task, taking the index of the task as the parameter. Currently, the level of abstraction above the TaskList, which is the Command classes, do checks that the index is valid, before calling these methods, thus, the TaskList class itself does not check the index's validity again. To check that these methods are never called with invalid indexes, let's add assert statements to each of these methods.

weijuey committed:
Restructure code that had bad readability, standardise exception handling, delete now defunct Ui class

weijuey committed:
Merge pull request #1 from weijuey/branch-assertions

Branch assertions

weijuey committed:
Merge branch 'master' of https://github.com/weijuey/ip

weijuey committed:
Merge branch 'master' into branch-A-CodeQuality

weijuey committed:
Merge pull request #2 from weijuey/branch-A-CodeQuality

Branch a code quality

weijuey committed:
Merge branch 'master' of https://github.com/weijuey/ip into branch-A-CodeQuality

weijuey committed:
Add .yml file for CI

weijuey committed:
Revert "Add .yml file for CI"

This reverts commit 741062562bc31292d8e3e1a9b53ea4cef30a496b.

weijuey committed:
Modify Task to implement Comparable interface so that there is a natural order of tasks

Tasks in the tasklist are currently always displayed in the order that they are added. This may not be the most useful way to show tasks to the user.

One obvious ordering of tasks that will be of interest to the user is chronological. Let's implement a natural order within Task classes that uses their given LocalDateTime attribute as the point of comparison. In addition, the TaskList class will implement a sort method that will allow other ways of sorting in the future.

weijuey committed:
Move method for getting a task's string format for saving to file from commands to the tasks

weijuey committed:
Create sort command

weijuey committed:
Fix up style issues

weijuey committed:
Merge branch 'branch-c-sort'

weijuey committed:
Rework layout of dialog boxes for more space for text

weijuey committed:
Rename sort options' internal names to match actual keyword used

weijuey committed:
Include correct date/time format to be entered when user enters it wrongly

weijuey committed:
Create user guide and fix some formatting

weijuey committed:
Add picture of Ui

weijuey committed:
Update Ume's personality

weijuey committed:
Merge branch 'branch-A-CodeQuality'

weijuey committed:
Create utility classes for testing purposes

weijuey committed:
Fix test tasks and implemented tests for find and sort commands
